_source/_posts/2020-07-02-spring-boot-using-java-modules.md

```diff
@@ -0,0 +1,449 @@
+okta.oauth2.clientSecret={clientSecret} 
+```
+
+You can find {clientId} and {clientSecret} in your application setup:  
+
+{% img blog/spring-boot-with-modules/okta-app.png alt:"Okta app setup" width:"800" %}{: .center-image }
+
+The {orgUrl} you can find on Okta dashboard:  
+
+{% img blog/spring-boot-with-modules/org-url.png alt:"Org Url" width:"800" %}{: .center-image }
+
+Now if you restart the app and navigate to `localhost:8080/bird` you'll get a login page appearing.
+
+## Using Java modules
+
+Now it is time to modularize the app. This is achieved by placing a file `module-info.java` on source root for each module. We'll be doing this for our two modules `application` and `persistence`. There are two ways to modularize a Java app: top-down and bottom-up. In this tutorial we'll be showing the bottom-up approach, that is modularizing the libraries before the app.
+
+### Modularize `persistence` library
+
+Create a module declaration file `persistence/src/main/java/module-info.java` with the following content:
+
+```java
+module com.okta.developer.modules.persistence {
+
+    requires java.annotation;
+    requires spring.beans;
+    requires spring.context;
+    requires spring.data.commons;
+    requires spring.data.mongodb;
+
+    exports com.okta.developer.animals.bird;
+
+}
+```
+
+Each `requires` keyword signalize that this module will be depending on some other module.
+Spring, on version 5, is not modularized yet so its JAR files doesn't have the `module-info.java`. 
+When you have a dependency on the `modulepath` (former classpath for non modular applications) like this they will be available as `automatic modules`.
+
+An `automatic module` gets its name from the property `Automatic-Module-Name` inside `MANIFEST.MF` or from the filename itself if that is absent.
+
+On Spring 5 the team was nice enough to put the `Automatic-Module-Name` for all the libraries. So those are the names used on our persistence app dependencies: `spring.beans`, `spring.context`, `spring.data.commons`, `spring.data.mongodb`.
+
+The `exports` keyword means that all classes on that package will be available for modules that uses this one.
+
+### Modularize `application` app
+
+Create a module declaration file `application/src/main/java/module-info.java` with the following content:
+
+```java
+module com.okta.developer.modules.app {
+
+    requires com.okta.developer.modules.persistence;
+
+    requires spring.web;
+    requires spring.boot;
+    requires spring.boot.autoconfigure;
+
+}
+```
+
+This one is similar to the first one but besides the Spring dependencies we also have the `com.okta.developer.modules.persistence` dependency. That is the other module we have developed.
+
+## Running the app
+
+Go to the project root and run 
+```bash
+mvn install && mvn spring-boot:run -pl application
+```
+
+Again, if everything went correctly you'll be able, after logging in, to navigate to `localhost:8080/bird` and see a JSON output.
+
```

Comment:
Add a "wrap up" section. similar to something like this: https://developer.okta.com/blog/2020/06/26/spring-boot-vue-kotlin#moving-forward-with-okta-vue-and-spring-boot

idea: Consider adding some sort of "private" utility class to show how that class cannot be used outside of it's module unless it is "exported", or maybe just show the compile error if you didn't export `com.okta.developer.animals.bird` 

(i'm not sure how that would flow the reader though, just ideas 😄)